### PR TITLE
Check the PK of the source of an error before closing chans from it

### PR DESCRIPTION
**Pull request by TheBlueMatt**
When we receive an error message from a peer, it can indicate a
channel which we should close. However, we previously did not
check that the counterparty who sends us such a message is the
counterparty with whom we have the channel, allowing any
connected peer to make us force-close any channel we have as long
as they know the channel id.

This commit simply changes the force-close logic to check that the
sender matches the channel's counterparty node_id, though as noted
in #105, we eventually need to change the indexing anyway to allow
absurdly terrible peers to open channels with us.

Found during review of #777.

Plus threw in a fix for docs on peer handler which I found while writing a SocketDescriptor wrapper in Java NIO.

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -3479,7 +3489,7 @@ impl<ChanSigner: ChannelKeys, M: Deref + Sync + Send, T: Deref + Sync + Send, K:
 			}
 		} else {
```
Comment (ariard) [inline lightning/src/ln/channelmanager.rs:3490]: About the zero-value channel id branch, wouldn't be better to sanitize early that case at `open_channel` reception ? Such value would be a hint of either our CSRNG being burst of the counterparty being buggy or sneaky. Either way we should discard that value from our internal channel map ?

**Review by ariard — COMMENTED** [review]
I'm sad to hear about this bug. Given that channel id are made of the `funding_txid` & `funding_output_index`, a peer could have reasonably guessed our channel id by observing onchain LN transactions...

Thinking further, we should modify the spec to blind channel id with a secret only known by channel counterparties, maybe even scoping `short_chan_id` for private channels ?

**Comment by TheBlueMatt** [comment]
> Thinking further, we should modify the spec to blind channel id with a secret only known by channel counterparties, maybe even scoping short_chan_id for private channels ?

We should just implement #105 - it'll solve this bug and also we can do per-channel locking at the same time.

File: lightning/src/ln/channelmanager.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3490]: Hmm? Not sure exactly what you mean - zero-value channel id means "close all channels with me", its not a specific reference to a channel.

**Comment by TheBlueMatt** [comment]
Added a test, as requested by @ariard on IRC, though I didn't remove the duplicate-closure on `[0; 32]` errors resulting from the PeerHandler calling `peer_disconnected(no_connection_possible: true)` after calling `handle_error` - we should behave correctly whether users have a custom `MessageHandler` or a custom `PeerHandler`.

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -3474,12 +3484,12 @@ impl<ChanSigner: ChannelKeys, M: Deref + Sync + Send, T: Deref + Sync + Send, K:
 			for chan in self.list_channels() {
 				if chan.remote_network_id == *counterparty_node_id {
```
Comment (naumenkogs) [inline lightning/src/ln/channelmanager.rs:3485]: Why you don't call `force_close..` with Some(*counterparty_node_id)?

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3485]: Doesn't matter much, but, sure, probably better for once we index by peers anyway.

**Comment by ariard** [comment]
Code Review ACK 97d6d91 modulo commit squash. Thanks for adding test!

**Comment by naumenkogs** [comment]
Two typos (f, isntead) in the commit name ` f call direct to self.force_close_channel_with_peer isntead `?

File: lightning/src/ln/peer_handler.rs [hunk 3]
```diff
@@ -88,10 +88,8 @@ pub trait SocketDescriptor : cmp::Eq + hash::Hash + Clone {
 }
 
 /// Error for PeerManager errors. If you get one of these, you must disconnect the socket and
-/// generate no further read_event/write_buffer_space_avail calls for the descriptor, only
-/// triggering a single socket_disconnected call (unless it was provided in response to a
-/// new_*_connection event, in which case no such socket_disconnected() must be called and the
-/// socket silently disconencted).
+/// generate no further read_event/write_buffer_space_avail/doscket_disconnected calls for the
```
Comment (naumenkogs) [inline lightning/src/ln/peer_handler.rs:91]: Typo "doscket_disconnected"

**Comment by TheBlueMatt** [comment]
Fixed the typo. Note that the "f..." commits, generally are intended to be squashed into previous commits before merge, they exist just to aid reviewers in seeing changes between revisions.

**Comment by TheBlueMatt** [comment]
Went ahead and squashed, only code change was the doc typo.

**Comment by ariard** [comment]
re-Code Review ACK bd8382a, changes are typo/squash

**Pull request merged**